File: src/main/scala/com/vc/medium/L875.scala
package com.vc.medium

object L875 {
  def minEatingSpeed(piles: Array[Int], H: Int): Int = {
    val n = piles.length
    def canEat(eat: Int): Boolean = {
      var countHours = 0
      piles.foreach(pile => {
        countHours += pile / eat
        if(pile % eat != 0) countHours += 1
      })
      countHours <= H
    }
    var end = piles.max
    var start = 1
    while(start <= end) {
      val mid = start + (end - start) / 2
      if(canEat(mid)) end = mid - 1
      else start = mid + 1
    }
    start
  }
}

File: src/main/java/com/vc/medium/L875.java
package com.vc.medium;

class L875 {
    public int minEatingSpeed(int[] piles, int H) {
        long lo = 1;
        long hi = 0;
        for(int i = 0; i < piles.length; i++) {
            hi += piles[i];
        }

        /**
         *  Minimize the maximum
         **/
        while(lo < hi) {
            long mid = lo + (hi - lo) / 2;
            boolean isPossible = isPossible(piles, mid, H);
            //System.out.println("Mid: "+mid+" isPossible: "+isPossible);
            if(!isPossible) lo = mid + 1;
            else hi = mid;
        }
        return (int)lo;
    }

    private boolean isPossible(int[] piles, long mid, int H) {
        int count = 0;
        for(int i = 0; i < piles.length; i++) {
            count += piles[i] / mid;
            if(piles[i] % mid != 0) count += 1;
            if(count > H) return false;
        }
        return true;
    }
}